### PR TITLE
 LoongArch: Add new addressing type ADDRESS_REG_REG support. 

### PR DESCRIPTION
**Pull request by chenglulu326**

### CODE REVIEW

**Comment by chenglulu326** [comment]
bot test this please

File: gcc/config/loongarch/constraints.md [hunk 1]
```diff
@@ -30,7 +30,7 @@
 ;; "h" "A constant call plt address."
 ;; "i" "Matches a general integer constant." (Global non-architectural)
 ;; "j" SIBCALL_REGS
-;; "k" <-----unused
+;; "k" "A memory operand whose address is formed by a base register and offset register."
```
Comment (xen0n) [inline gcc/config/loongarch/constraints.md:33]: Not the same as the description below?

File: gcc/config/loongarch/constraints.md [hunk 2]
```diff
@@ -113,6 +113,13 @@
 (define_register_constraint "j" "SIBCALL_REGS"
   "@internal")
 
+(define_memory_constraint "k"
+  "A memory operand whose address is formed by a base register and offset.
+   And offset is register or 12 bit immediate."
+  (and (match_code "mem")
+       (not (match_test "loongarch_14bit_shifted_offset_address_p (XEXP (op, 0), mode)"))
+       (not (match_test "loongarch_12bit_offset_address_p (XEXP (op, 0), mode)"))))
```
Comment (xen0n) [inline gcc/config/loongarch/constraints.md:121]: Judging from the code below this does seem like intended for `ldx/stx`, and not `alsl`, and I don't get why there's need to restrict the range to 14bit and 12bit.

File: gcc/config/loongarch/constraints.md (hunk 1 above)
Comment (xen0n) [reply gcc/config/loongarch/constraints.md:33]: Okay for `ldx/stx` it's of course `reg+reg` and this sentence is correct.

File: gcc/config/loongarch/loongarch.cc [hunk 3]
```diff
@@ -3258,6 +3309,71 @@ loongarch_split_move_insn (rtx dest, rtx src, rtx insn)
...
+      },
+      {
+	"ldx.bu\t%%0,%%1",
+	"ldx.hu\t%%0,%%1",
+	"ldx.w\t%%0,%%1",
+	"ldx.d\t%%0,%%1",
+      }
+    };
+  snprintf (buf, sizeof (buf), insn[ldr][index]);
+  return buf;
```
Comment (xen0n) [inline gcc/config/loongarch/loongarch.cc:3343]: why not directly return the pointer to the const string? it seems there's no postprocessing of any kind so the `snprintf` call seems superfluous.

**Comment by chenglulu326** [comment]
bot test this please

**Pull request closed without merging**